enzyme/Enzyme/MLIR/Interfaces/GradientUtilsReverse.cpp

```diff
@@ -0,0 +1,405 @@
+  SmallVector<OpBuilder *> builders = {&initBuilder, &revBuilder};
+  if (auto iface = dyn_cast<AutoDiffOpInterfaceReverse>(op)) {
+    return iface.clearGradient(builders, this);
+  }
+}
+
+bool mlir::enzyme::MGradientUtilsReverse::hasInvertPointer(mlir::Value v){
+  return (invertedPointers.contains(v)) || (invertedPointersGlobal.contains(v));
+}
+
+bool MGradientUtilsReverse::visitChildCustom(Operation * op, OpBuilder &builder){
+  std::string nameDiffe = "diffe_" + op->getName().getDialectNamespace().str() + "_" + op->getName().stripDialect().str();
+  std::string nameStore = "store_" + op->getName().getDialectNamespace().str() + "_" + op->getName().stripDialect().str();
+
+  StringRef srDiffe(nameDiffe);
+  StringRef srStore(nameStore);
+  
+  OperationName opNameDiffe(srDiffe, op->getContext());
+  OperationName opNameStore(srStore, op->getContext());
+
+  Operation * symbolDiffe = symbolTable.lookupNearestSymbolFrom(op, opNameDiffe.getIdentifier());
+  Operation * symbolStore = symbolTable.lookupNearestSymbolFrom(op, opNameStore.getIdentifier());
+  
+  if (symbolDiffe != nullptr){
+    SmallVector<Value> caches;
+    if(symbolStore != nullptr){
+      Operation * newOp = getNewFromOriginal(op);
+
+      func::FuncOp funcStore = cast<func::FuncOp>(symbolStore);
+      
+      SmallVector<Type, 2> storeResultTypes;
+      for (auto x : funcStore.getFunctionType().getResults()){
+        storeResultTypes.push_back(x);
+      }
+
+      SmallVector<Value, 2> storeArgs;
+      for (auto x : newOp->getOperands()){
+        storeArgs.push_back(x);
+      }
+
+      OpBuilder storeBuilder(newOp);
+      func::CallOp storeCI = storeBuilder.create<func::CallOp>(op->getLoc(), srStore, storeResultTypes, storeArgs);
+      for (auto x : storeCI.getResults()){
+        caches.push_back(cacheForReverse(x, storeBuilder));
+      }
+    }
+    
+    SmallVector<Value, 2> args;
+    
+    bool hasGradient = false;
+    for (Value opResult : op->getResults()){
+      if(hasInvertPointer(opResult)){
+        hasGradient = true;
+        Value invertValue = invertPointerM(opResult, builder);
+        args.push_back(invertValue);
+      }
+    }
+    for (Value cache : caches){
+      args.push_back(popCache(cache, builder));
+    }
+
+    SmallVector<Type, 2> resultTypes;
+    for (auto x : op->getOperands()){
+      resultTypes.push_back(x.getType());
+    }
+
+    func::CallOp dCI = builder.create<func::CallOp>(op->getLoc(), srDiffe, resultTypes, args);
+    for (int i = 0; i < op->getNumOperands(); i++){
+      mapInvertPointer(op->getOperand(i), dCI.getResult(i), builder);
+    }
+
+    return true;
+  }
+  return false;
+}
+
+LogicalResult MGradientUtilsReverse::visitChildReverse(Operation *op, OpBuilder &builder) {
+  if (mode == DerivativeMode::ReverseModeGradient) {
+    if (auto binst = dyn_cast<BranchOpInterface>(op)) {
+      
+    }
+    else if (auto binst = dyn_cast<func::ReturnOp>(op)) {
```

Comment:
We probably want to handle all return-like operations, not just `func::ReturnOp`. Something like `RegionBranchOpInterface` would be more appropriate.